### ISSUE TITLE
The WebGL example never finishes loading

### ISSUE THREAD

**Issue by antocuni** [pyscript/pyscript#678]
The WebGL example shows a very weird behavior, which was discovered by #676 .
The relevant test in the PR is this:
https://github.com/pyscript/pyscript/blob/b90813bed54787ea14b074d58cadc4d17f0e4719/pyscriptjs/tests/test_zz_examples.py#L200-L206

The pytest log is the following (some lines deleted for better readability):
```
[  0.00 page.goto      ] examples/webgl/raycaster/index.html
[  0.01 http_server    ] "GET /examples/webgl/raycaster/index.html HTTP/1.1" 200 -
[  0.02 http_server    ] "GET /examples/webgl/raycaster/style.css HTTP/1.1" 200 -
[  0.02 http_server    ] "GET /build/pyscript.css HTTP/1.1" 200 -
[  0.02 http_server    ] "GET /build/pyscript.js HTTP/1.1" 200 -
...cut...
[  0.28 console.log    ] RUNTIME READY
[  0.28 console.log    ] config set!
[  0.28 console.log    ] initializers set
[  0.28 console.log    ] post initializers set
[  0.28 console.log    ] post initializers set
[  0.28 console.log    ] post initializers set
[  0.28 console.log    ] connected
[  0.28 console.log    ] config set!
[  0.28 console.log    ] config set {autoclose_loader: true, runtimes: Array(1)}
[  0.28 console.log    ] Initializing runtimes...
[  0.28 console.error  ] Failed to load resource: net::ERR_CONNECTION_REFUSED
[  0.38 console.log    ] creating pyodide runtime
[  2.82 console.log    ] Python initialization complete
[  2.82 console.log    ] loading micropip
[  2.82 console.log    ] Loading micropip, pyparsing, packaging, distutils
[  3.15 console.log    ] Loaded micropip, pyparsing, packaging, distutils
[  3.15 console.log    ] loading pyscript...
[  3.88 console.log    ] done setting up environment
[  3.88 console.log    ] RUNTIME READY
[  3.88 console.log    ] Collecting nodes to be mounted into python namespace...
[  3.88 console.log    ] evaluate
[  3.89 console.log    ] ----> changed out to py-bd3390f1-3834-cbd7-da68-0caf534ffb9a true
[  3.89 console.log    ] ----> changed out to py-bd3390f1-3834-cbd7-da68-0caf534ffb9a true
[  3.96 console.log    ] THREE.WebGLRenderer 89
[  4.39 console.warning] [.WebGL-0xc400345b00]GL Driver Message (OpenGL, Performance, GL_CLOSE_PATH_NV, High): GPU stall due to ReadPixels
[  4.44 console.warning] [.WebGL-0xc400345b00]GL Driver Message (OpenGL, Performance, GL_CLOSE_PATH_NV, High): GPU stall due to ReadPixels
[  4.49 console.warning] [.WebGL-0xc400345b00]GL Driver Message (OpenGL, Performance, GL_CLOSE_PATH_NV, High): GPU stall due to ReadPixels
[  4.53 console.warning] [.WebGL-0xc400345b00]GL Driver Message (OpenGL, Performance, GL_CLOSE_PATH_NV, High): GPU stall due to ReadPixels (this message will no longer repeat)
```

After that, it never reaches this point, and thus it never prints `===PyScript page fully initialized===`:
https://github.com/pyscript/pyscript/blob/b90813bed54787ea14b074d58cadc4d17f0e4719/pyscriptjs/src/components/pyconfig.ts#L119-L122

I suspect it's because of the `while True:` loop in the `<py-script>`, but we should probably sit down and decide whether it's the behavior that we want.

Moreover, another weird thing is that if I run the example manually, the loader screen correctly hides, but if I run it with `playwright` inside the tests, the loader stays forever.

**Comment by sumahadevan** [comment]
A similar issue is seen on Ubuntu 22.04 when running another example Panel Streaming Demo, panel_stream.html, on localhost:8080 (after first applying the fix for resolving the pyscript.js path as indicated in #688). Namely, the page displays correctly but the loader screen stays on screen forever.

The same also happens when the example is run by simply double clicking directly on panel_stream.html (which is possible because the code is all client-side).

I am not sure if I should log a separate issue for this new example or it is a common issue. The new testing machinery may help to clarify this, and I would be happy to help flesh out some of the tests to identify the cause of the problem(s).

**Comment by sumahadevan** [comment]
I found one common factor between the two examples which don't finish loading - WebGL and Panel Streaming. Both of them execute a fourth step 'Initializing scripts...' as the last initialization step. On the other hand, examples like Simple Static Bokeh Plot or Simple D3 Visualization which DO finish loading only execute the first three steps (Loading runtime..., Runtime created..., Initializing components...). Of course this hypothesis needs to be confirmed by debugging the code.

**Comment by fpliger** [comment]
I think the core issue here is the quick and dirty implementation of the loader, the sequence of events,  and how we are managing the loader itself. We need to better manage event hooks and closing the loader (and have a less invasive loader).

**Comment by sumahadevan** [comment]
@antocuni This issue is resolved by the same fix as #751, and the fix is in fact the same that you have suggested in the discussion above, namely WE SHOULD NOT USE "await" before invoking "script.evaluate()" within the scripts-processing loop in runtime.ts. I made this one-line change and verified that the above issue is indeed resolved, namely the WebGL example successfully completes loading. Ths same fix also resolves #751 related to the asynchronous behavior of the pyscript tags.

The reason for facing this issue is the following - the await WILL BLOCK all following code in the function is it used in, but WILL NOT BLOCK the calling function one level above. So putting an await directly in the loop means that all following scripts in the same loop are blocked (which we dont want), while having the await at a much lower level (for example: await runtime.run) means that the following scripts in the loop, AND higher-level calling functions, can execute.

Secondly, despite having an infinite loop in our code, all the messages on screen are now displayed AND cleared without getting blocked. The infinite loop in the WebGL Python code is really no different from the use of an infinite loop on a Canvas element in a HTML game, and does not affect (non-blocking) code that executes before the infinite loop actually starts - this non-blocking code that executes before the loop starts will complete successfully.

The following Mozilla Developer Network (MDN) links throw light on the behaviour of async and await in Javascript code (in particular, the fact that the position of the await in the code is crucial!).

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

**Comment by tedpatrick** [comment]
With the changes in script.evaluate() in #796, this is now resolved. I have removed the XFAIL from the test in https://github.com/pyscript/pyscript/pull/829 and it is passing consistently. Also in the 2022.06.1 release, the example is working correctly.

https://pyscript.net/examples/webgl/raycaster/index.html

Closing issue
